minetest/minetest
# [NO SQUASH] Settings: Proper priority hierarchy

The goal of this PR is to provide clear settings priorities. There have been various issues regarding the somewhat confusing structure, hence this is an attempt at solving it.

All known and persistent Settings objects are now held in a static array. Getter functions in `Settings` move to index 0 as fallback (if there was no such setting).

## To do

This PR is Ready for review


- [x] re-enable mapgen settings unittest
- [x] delete game setting after leaving
- [x] perhaps replace the array-based approach with a linked list
  * eases the parent getter function
- [x] Fix map_meta.txt corrupting for no reason (end tag and garbage placed too early)

## How to test

1. run the unittests
2. Backup an existing world and join it afterwards
    * new map data must generated as expected
3. Mess around in the settings
4. Create a new world and enable a Lua mapgen mod. It must still work as usual.
5. Create/recycle a world with any mapgen
   * Join, leave
   * Change `mg_name` to `valleys` or `v5`
   * Rejoin and leave. The new settings are appended as expected
6. `FlagDesc` tests: https://github.com/minetest/minetest/pull/10640#issuecomment-731725845
